geoprocessing/src/modules/shapefiles/shapefiles.service.ts

```diff
@@ -1,33 +1,68 @@
-import { Injectable, Logger } from '@nestjs/common';
+import { BadRequestException, Injectable, Logger } from '@nestjs/common';
+import { FeatureCollection } from 'geojson';
 import { FileService } from '../files/files.service';
+
 const mapshaper = require('mapshaper');
 
+const MIN_REQUIRED_FILES = ['prj', 'dbf', 'shx', 'shp'];
 @Injectable()
 export class ShapefileService {
   private readonly logger: Logger = new Logger(ShapefileService.name);
   constructor(private readonly fileService: FileService) {}
 
   private async shapeFileToGeoJson(fileInfo: Express.Multer.File) {
+    if (
+      !(await this.fileService.areFilesInFolder(
+        fileInfo.path.replace('.zip', ''),
+        MIN_REQUIRED_FILES,
+      ))
+    ) {
+      throw new Error('Some required file mising');
+    }
+    console.log('works');
     const outputKey = `shapefile-${new Date().getTime()}.geojson`;
 
     const _geoJson = await mapshaper.applyCommands(
-      `-i ${fileInfo.path.replace('.zip', '')}/*.shp -info -o ${outputKey}`,
+      `-i no-topology snap ${fileInfo.path.replace(
+        '.zip',
+        '',
+      )}/*.shp -proj EPSG:4326 -clean rewind -info -o ${outputKey}`,
     );
 
     return JSON.parse(_geoJson[outputKey].toString('utf-8'));
   }
 
+  isValidGeoJson(geoJson: FeatureCollection): Error | void {
```

Comment:
Thanks @hotzevzl . It seemed nasty to explicitly throw an error inside a try/catch block for me at first.
591e7097cd28b32cb3cbe8effd61644ad56dd28b

PS: Now wondering if we should add any error that may occur here to `BadRequestException` message...

Reply:
up to you - I'd say where it is easy and meaningful to send useful error details we should do it, especially if the situation may be due to user error (which could be common, for example an oversight) rather than application error (such as an bug in the way a specific software/library assembles a shapefile), as the user may not know what to do in any case in the latter case.